gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Channelizer.java

```diff
@@ -123,7 +124,9 @@ protected void initChannel(final SocketChannel socketChannel) throws Exception {
             }
 
             if (sslCtx.isPresent()) {
-                pipeline.addLast(sslCtx.get().newHandler(socketChannel.alloc(), connection.getUri().getHost(), connection.getUri().getPort()));
+                SslHandler sslHandler = sslCtx.get().newHandler(socketChannel.alloc(), connection.getUri().getHost(), connection.getUri().getPort());
+                sslHandler.setHandshakeTimeoutMillis(0);
```

Comment:
I recently noticed an exception that is being thrown from the SslHandler when setting this value to 0.

`[ERROR] org.apache.tinkerpop.gremlin.driver.Handler$GremlinResponseHandler - Could not process the response
io.netty.handler.ssl.SslClosedEngineException: SSLEngine closed already
	at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:859)
	at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:798)
	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:779)
	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:1971)
	at io.netty.handler.ssl.SslHandler.closeOutboundAndChannel(SslHandler.java:1940)
	at io.netty.handler.ssl.SslHandler.close(SslHandler.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:622)
	at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:606)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.close(CombinedChannelDuplexHandler.java:505)
	at io.netty.channel.ChannelOutboundHandlerAdapter.close(ChannelOutboundHandlerAdapter.java:77)
	at io.netty.channel.CombinedChannelDuplexHandler.close(CombinedChannelDuplexHandler.java:316)
	at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:622)
	at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:606)
	at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:472)
	at io.netty.handler.codec.http.websocketx.WebSocketClientProtocolHandshakeHandler$2.run(WebSocketClientProtocolHandshakeHandler.java:122)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:829)`

I've tried several workarounds to get rid of this issue with closing the SslHandler. These workarounds include subclassing SslHandler to modify how close works if the handshake times out, removing the SslHandler from the pipeline and swallowing the exception if the handshake failed. However, all of these workarounds have drawbacks so instead I propose that we go back to your initial recommendation which is to make the SSL handshake timeout a percentage of `connectionSetupTimeoutMillis`. I feel like 0.9*connectionSetupTimeoutMillis would still enable configurability for this value while preventing any timing issues with the WebSocket handshake timeout.

What are you thoughts on this @divijvaidya ? Thanks.

Reply:
Please tell me more. I don't think I understood the cause of the exception above.

From the above exception stack trace, it looks like when we try to close SslHandler (during termination of a connection), it is already closed and hence, throws an exception. Why is it already closed? What does setting the timeout to 0 vs. say 100 has to do with it?

Reply:
That "SSLEngine closed already" exception is just a general error message because the handshake future wasn't given a specific cause. When updating the handshake future with an SSLHandshakeException, the stack trace looks like

`[ERROR] org.apache.tinkerpop.gremlin.driver.Handler$GremlinResponseHandler - Could not process the response
javax.net.ssl.SSLHandshakeException: SSL handshake timed out.
	at org.apache.tinkerpop.gremlin.driver.handler.WebSocketClientHandler.userEventTriggered(WebSocketClientHandler.java:127)
	at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:332)
	at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:324)
	at io.netty.handler.codec.http.websocketx.WebSocketClientProtocolHandshakeHandler$2.run(WebSocketClientProtocolHandshakeHandler.java:121)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:829)`

For more information, this is what the 3.5.4 driver logs show:
`[ERROR] org.apache.tinkerpop.gremlin.driver.Handler$GremlinResponseHandler - Could not process the response
io.netty.handler.ssl.SslHandshakeTimeoutException: handshake timed out after 10000ms
	at io.netty.handler.ssl.SslHandler$7.run(SslHandler.java:2113)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:829)`

If you think this new stack trace is fine then we can keep the default SSL handshake timeout as 0. Otherwise, it may be cleaner to just use a percentage of connectionSetupTimeoutMillis. And just to be clear, this is just an exception that is logged in logs. The user won't see this exception as it is part of the close/cleanup process happening in the background.

Reply:
I think that the new stack trace i.e. `[ERROR] org.apache.tinkerpop.gremlin.driver.Handler$GremlinResponseHandler - Could not process the response javax.net.ssl.SSLHandshakeException: SSL handshake timed out` is fine as it provides the reason for connection failure i.e. ssl timeout.

I am saying this because the alternative to go for a %age based timeout adds another knob that the users have to tune and they don't have good heuristics to tune it since the %age is a function of network weather at that moment. That is the reason I have an inclination to avoid adding a new configuration or adding a %age based solution.